src/languageclient.rs

```diff
@@ -2202,16 +2202,42 @@ impl State {
         let user_data: VimCompleteItemUserData = serde_json::from_str(&user_data)?;
 
         let mut edits = vec![];
+        let mut cur_line = line + 1;
+        let mut cur_char = character + 1;
         if let Some(edit) = user_data.text_edit {
-            self.command("undo")?;
-            edits.push(edit.clone());
+            let word_len = completed_item.word.len();
+            let start : usize  = &(character as usize) - word_len;
+            let end : usize = character as usize;
+            let mut current_line: String = self.call(None, "getline", json!([line+1]))?;
+            current_line.replace_range(start..end, "");
+
+            if self.call::<_, i64>(None, "setline", json!([line+1, current_line]))? != 0 {
+                bail!("Failed to set buffer content!");
+            }
+            let mut edit_clone = edit.clone();
+            edit_clone.range.start.character = start as u64;
+            edit_clone.range.end.character = start as u64;
```

Comment:
This is really wired. Is this necessary or even working?

Reply:
yes for both.  What's going on here is:

1. Removing the word inserted by base autocomplete behavior.  (replacing the "undo" command)
2. Editing the start/end position of the ranges from the given completion response. (the completion response is cached by completion managers, and often does not have the correct position according to the current cursor position).

I talk more about them below.